core/trino-main/src/main/java/io/trino/operator/TopNProcessor.java

```diff
@@ -76,30 +87,46 @@ public void addInput(Page page)
 
     public Page getOutput()
     {
-        if (outputIterator == null) {
+        if (outputIterator == null && topNBuilder != null) {
             // start flushing
             outputIterator = topNBuilder.buildResult();
         }
 
-        Page output = null;
-        if (outputIterator.hasNext()) {
+        Page output;
+        if (outputIterator != null && outputIterator.hasNext()) {
             output = outputIterator.next();
         }
         else {
-            outputIterator = emptyIterator();
+            closeTopNBuilder();
+            return null;
         }
         updateMemoryReservation();
         return output;
```

Comment:
The updated logic flow is a quite hard to understand. I think something like this would make the invariants at each step easier to follow:
```java
        if (outputIterator == null) {
            // start flushing
            outputIterator = topNBuilder == null ? emptyIterator() : topNBuilder.buildResult();
            topNBuilder = null;
        }

        Page output = null;
        if (outputIterator.hasNext()) {
            output = outputIterator.next();
        }
        else {
            outputIterator = null;
        }
        updateMemoryReservation();
        return output;
```
I believe this does what you want.

Reply:
LGTM except the timing of "topNBuilder = null;". topNBuilder shouldn't be nullified until outputIterator is nullified. Otherwise there will be an error for memory accounting.